src/libstd/sync/remutex.rs

```diff
@@ -0,0 +1,248 @@
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use prelude::v1::*;
+
+use cell::UnsafeCell;
+use marker;
+use ops::{Deref, DerefMut};
+use sync::poison::{self, TryLockError, TryLockResult, LockResult};
+use sys::mutex as sys;
+use fmt;
+
+/// A re-entrant mutual exclusion
+///
+/// This mutex will block *other* threads waiting for the lock to become available. The thread
+/// which has already locked the mutex can lock it multiple times without blocking, preventing a
+/// common source of deadlocks.
+///
+/// # Examples
+///
+/// ```
+/// use std::sync::ReentrantMutex;
+///
+/// fn locking_mod(data: ReentrantMutex<i32>) {
+///     let mut data = data.lock().unwrap();
+///     *data += 1;
+/// }
+///
+/// let data = ReentrantMutex::new(0);
+/// {
+/// // We take the lock twice here. It won’t deadlock.
+/// let lock = data.lock();
+/// locking_mod(data);
+/// }
+/// ```
+#[unstable(feature = "reentrant_mutex", reason = "new API")]
+pub struct ReentrantMutex<T> {
+    inner: Box<sys::ReentrantMutex>,
+    poison: poison::Flag,
+    data: UnsafeCell<T>,
+}
+
+unsafe impl<T: Send> Send for ReentrantMutex<T> {}
+unsafe impl<T: Send> Sync for ReentrantMutex<T> {}
+
+
+/// An RAII implementation of a "scoped lock" of a mutex. When this structure is
+/// dropped (falls out of scope), the lock will be unlocked.
+///
+/// The data protected by the mutex can be accessed through this guard via its
+/// Deref and DerefMut implementations
+#[must_use]
+#[unstable(feature = "reentrant_mutex", reason = "new API")]
+pub struct ReentrantMutexGuard<'a, T: 'a> {
+    // funny underscores due to how Deref/DerefMut currently work (they
+    // disregard field privacy).
+    __lock: &'a ReentrantMutex<T>,
+    __poison: poison::Guard,
+}
+
+impl<'a, T> !marker::Send for ReentrantMutexGuard<'a, T> {}
+
+
+#[unstable(feature = "reentrant_mutex", reason = "new API")]
+impl<T> ReentrantMutex<T> {
+    /// Creates a new reentrant mutex in an unlocked state.
+    pub fn new(t: T) -> ReentrantMutex<T> {
+        ReentrantMutex {
+            inner: box unsafe { sys::ReentrantMutex::new() },
+            poison: poison::FLAG_INIT,
+            data: UnsafeCell::new(t),
+        }
+    }
+
+    /// Acquires a mutex, blocking the current thread until it is able to do so.
+    ///
+    /// This function will block the caller until it is available to acquire the mutex.
+    /// Upon returning, the thread is the only thread with the mutex held. When the thread
+    /// calling this method already holds the lock, the call shall succeed without
+    /// blocking.
+    ///
+    /// # Failure
+    ///
+    /// If another user of this mutex panicked while holding the mutex, then
+    /// this call will return failure if the mutex would otherwise be
+    /// acquired.
+    pub fn lock(&self) -> LockResult<ReentrantMutexGuard<T>> {
```

Comment:
cc @aturon, this is a bit of an interesting case for poisoning. I believe that to be safe a reentrant mutex can only allow `Deref` access to its internals (not `DerefMut`, see below), so there must be some other form of interior mutability. This raises a question in my mind of whether we need to handle poisoned data here in that case. I think we still do because you can still see the poisoned data in other threads in normal code, just something I found interesting.

To be clear though I don't mind landing this either way, that's what `#[unstable]` is for!